### ISSUE TITLE
Explicit Motion Timing Protocols and Current Controllers

### ISSUE THREAD

**Issue by sjkelly** [RAPcores/rapcores#153]
With the microstepping work on main we now have the ability to dynamically alter currents as part of motion planning. The current DDA could be expanded to support these modes. An alternative that may dovetail better with other firmwares and frameworks would be an explicit timing mode (whereas DDA is an implicit form). We would need to specify:

- Duration
- Clock ticks/step
- Delta clock ticks/step
- Initial current
- Delta current/step 

Wild Speculation:

The main hypothesis is that we are better served using high level dynamics modeling to regulate current (as an analogue for joint torques). Part of this is that the state space for torque control on the motor is two dimensional (rotation and vector length), making a simple PID controller inadequate. PID is sufficient for isolated cases, but we know from low-cost servo options that PID suffers from instability at given velocity windows. 

For BLDC this issue is alleviated to some extent by driving the motor at high RPM with high gear ratios. Steppers cannot do this. Moreover, FOC on a stepper has far fewer encoder ticks per electrical cycle (or phase) than a BLDC. This means the error term derived from the encoder is far lower resolution than the number of divisions of our space vector (e.g 5 ticks vs 64 microsteps).

That is not to say controllers in rapcores is not possible or valuable. One approach in rapcores maybe to use a PID or other simple controller to alter the current vector if the encoder is showing lag in the system due to load as a last resort. We will still need an external controller to adjust the velocities and torques periodically (e.g control the angle element of the controls problem). But I think a far more powerful approach is to develop a state space from the torque data during a move to improve the dynamics models on the fly, obviating the need for a controller in rapcores. Given a 1khz update rate we will generate gigs of data that could be used to implement this type of controller.

Links:

- https://www.machinedesign.com/automation-iiot/article/21826706/20-minute-tuneup-zoning-in-on-position-loops#:~:text=In%20a%20series%2C%20or%20cascaded,calculate%20the%20corresponding%20following%20error.&text=The%20greater%20the%20position%20error,a%20direction%20reducing%20the%20error.
- https://www.machinedesign.com/automation-iiot/article/21828255/20-minute-tuneup-put-a-pid-on-it
- https://www.machinedesign.com/home/contact/21838863/motion-system-design

**Comment by sjkelly** [comment]
XRef #149 

Adding to 0.2 milestone

**Comment by sjkelly** [comment]
I'm was conflating two things here. Fixed point is not antithetical to current interpolation, so my reasoning is flawed. Also increasing the DDA base frequency would address my point of BLDC drive frequency as well. Upon more investigation, fixed point is the only way to achieve good precision. I think when I wrote this I was worried that LUT use of the DDA would prevent additional control parameters and I was looking at simpler examples. LUT use shouldn't be a hard constraint yet. Since this issue, we have documented some basic stepper motor models, so the controls intuition I originally wrote here is mostly correct, but it is a subset of wider control solutions. We may let the path planning software implement a control, or we can do onboard, both are valid. Nothing is concretely actionable here or a real concern anymore, so I am closing.
